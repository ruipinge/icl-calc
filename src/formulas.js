/*
function calcPosMeridian(sphere) {
    return -1.33756 + (0.9446 * sphere);
};

function calcNegMeridian(sphere, cylinder) {
    return -1.33756 + (0.9446 * (sphere + cylinder));
};

export const iclSphere = selector({
  key: 'iclSphere',
  get: ({get}) => {
    const val = get(patientState);
    const val = get(biometricsState),
      negMer = calcNegMeridian(val.sphere, val.cylinder);

    return `${val.name} (${val.dateOfBirth}), ${val.age} years old, ${val.eye} eye`;
    return Number((negMer).toFixed(2));
  }
});

export const iclCylinder = selector({
  key: 'iclCylinder',
  get: ({get}) => {
    const val = get(biometricsState),
      posMer = calcPosMeridian(val.sphere),
      negMer = calcPosMeridian(val.sphere, val.cylinder);

    return Number((posMer - negMer).toFixed(2));
  }

});

export const iclAxis = selector({
  key: 'iclAxis',
  get: ({get}) => {
    const val = get(biometricsState),
      axis = val.axis;

    // ICl axis
    if (axis >= 0 && axis <= 90) {
      return axis + 90;
    } else if (axis >=91 && axis <= 180) {
      // TODO: this looks like a bug
      return axis + 90;
    }

    return axis;
  }
 */

export default null;