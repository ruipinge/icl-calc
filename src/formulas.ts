const calcPosMeridian = (sphere: number) => {
    return -1.33756 + (0.9446 * sphere);
};

const calcNegMeridian = (sphere: number, cylinder: number) => {
    return -1.33756 + (0.9446 * (sphere + cylinder));
};

export const calcIclSphere = (sphere: number, cylinder: number) => {
    return calcNegMeridian(sphere, cylinder);
};

export const calcIclCylinder = (sphere: number, cylinder: number) => {
    const posMer = calcPosMeridian(sphere),
      negMer = calcNegMeridian(sphere, cylinder);

    return posMer - negMer;
};

export const calcIclAxis = (axis: number) => {
  if (axis >= 0 && axis < 90) {
    return axis + 90;
  }
  
  if (axis > 90 && axis <= 180) {
    return axis - 90;
  }
  
  return axis;
}

export const calcIclCylinderEquivalent = (sphere: number, cylinder: number) => {
  const negCylinder = calcNegMeridian(sphere, cylinder) - calcPosMeridian(sphere);
  return calcPosMeridian(sphere) + (negCylinder / 2);
}