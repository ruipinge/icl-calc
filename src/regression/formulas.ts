import { LENS_SIZES_MAP, LensSizeId } from '../matrix/data';

export interface RegressionInfo {
  acd: number;
  ata: number;
  clr: number;
  se: number;
  age: number;
}

interface RegressionExt {
  vaultPredictionVal0: number;
  probExpVal0: number;
  probExpVal1: number;
}

type LensSizeRegressionType = {
  [key in LensSizeId]: RegressionExt;
};

export const RegressionLensSizes: LensSizeRegressionType = {
  small: {
    vaultPredictionVal0: 4132.04 - 260.99,
    probExpVal0: -1.259,
    probExpVal1: -5.829
  },
  medium: {
    vaultPredictionVal0: 4132.04,
    probExpVal0: -1.401,
    probExpVal1: -5.733
  },
  large: {
    vaultPredictionVal0: 4132.04 + 330.19,
    probExpVal0: -2.191,
    probExpVal1: -3.775
  }
};

const vaultPredictionVal1 = (ri: RegressionInfo) =>
  -287.45 * ri.ata + -0.37 * ri.clr + -20.39 * ri.se + -4.82 * ri.age;

const compression = ({
  ata,
  lensSize
}: {
  ata: number;
  lensSize: number;
}): number => ata - lensSize;

const probExpFactor0 = ({
  ri,
  lensSize
}: {
  ri: RegressionInfo;
  lensSize: number;
}) =>
  0.028 * ri.age -
  0.506 * compression({ ata: ri.ata, lensSize: lensSize }) +
  0.005 * ri.clr +
  0.202 * ri.se;

const probExpFactor1 = ({
  ri,
  lensSize
}: {
  ri: RegressionInfo;
  lensSize: number;
}) =>
  -0.076 * ri.age +
  3.721 * compression({ ata: ri.ata, lensSize: lensSize }) +
  0.0 * ri.clr -
  0.164 * ri.se;

export const vaultPrediction = ({
  ri,
  lensSizeId
}: {
  ri: RegressionInfo;
  lensSizeId: LensSizeId;
}): number =>
  RegressionLensSizes[lensSizeId].vaultPredictionVal0 + vaultPredictionVal1(ri);

export const cornea2endothelium = ({
  ri,
  lensSizeId
}: {
  ri: RegressionInfo;
  lensSizeId: LensSizeId;
}): number => ri.acd - (220 + vaultPrediction({ ri, lensSizeId })) / 1000;

export const probability = ({
  ri,
  lensSizeId
}: {
  ri: RegressionInfo;
  lensSizeId: LensSizeId;
}) =>
  1 /
  (1 +
    Math.exp(
      RegressionLensSizes[lensSizeId].probExpVal0 +
        probExpFactor0({ ri: ri, lensSize: LENS_SIZES_MAP[lensSizeId].value })
    ) +
    Math.exp(
      RegressionLensSizes[lensSizeId].probExpVal1 +
        probExpFactor1({ ri: ri, lensSize: LENS_SIZES_MAP[lensSizeId].value })
    ));
