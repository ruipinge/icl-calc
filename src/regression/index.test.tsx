import { RI } from './formulas.test';
import { Regression } from '.';
import TestRenderer from 'react-test-renderer';

it('renders without crashing', () => {
  const tree = TestRenderer.create(<Regression {...RI} />).toJSON();
  expect(tree).toMatchSnapshot();
});
