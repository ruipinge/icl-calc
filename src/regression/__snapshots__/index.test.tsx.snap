// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <h4>
        Vault Prediction
      </h4>
      <table
        className="table table-bordered table-hover"
      >
        <thead>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              Lens Size
            </th>
            <th
              className="col-4 text-right"
              scope="col"
            >
              Vault (μm)
            </th>
            <th
              className="col-4 text-right"
              scope="col"
            >
              Corneal Endothelium to ICL (mm)
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              12.6 mm
            </th>
            <td
              className="col-4 text-right"
            >
              448.2
            </td>
            <td
              className="col-4 text-right"
            >
              2.74
            </td>
          </tr>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              13.2 mm
            </th>
            <td
              className="col-4 text-right"
            >
              709.19
            </td>
            <td
              className="col-4 text-right"
            >
              2.48
            </td>
          </tr>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              13.7 mm
            </th>
            <td
              className="col-4 text-right"
            >
              1039.38
            </td>
            <td
              className="col-4 text-right"
            >
              2.15
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <h4>
        Probability of 250 &lt; Vault &lt; 1000 (μm)
      </h4>
      <table
        className="table table-bordered table-hover"
      >
        <thead>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              Lens Size
            </th>
            <th
              className="col-4 text-right"
              scope="col"
            >
              Probability (%)
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              12.6 mm
            </th>
            <td
              className="col-4 text-right"
            >
              83.8
            </td>
          </tr>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              13.2 mm
            </th>
            <td
              className="col-4 text-right"
            >
              77.7
            </td>
          </tr>
          <tr>
            <th
              className="col-4"
              scope="col"
            >
              13.7 mm
            </th>
            <td
              className="col-4 text-right"
            >
              11.2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
]
`;
